aptos-move/framework/aptos-framework/sources/AptosGovernance.move

```diff
@@ -0,0 +1,360 @@
+    /// AptosGovernance.
+    public fun update_governance_config(
+        _proposal: GovernanceProposal,
+        min_voting_threshold: u128,
+        required_proposer_stake: u128,
+        voting_period_secs: u64,
+    ) acquires GovernanceConfig, GovernanceEvents {
+        let governance_config = borrow_global_mut<GovernanceConfig>(@AptosFramework);
+        governance_config.voting_period_secs = voting_period_secs;
+        governance_config.min_voting_threshold = min_voting_threshold;
+        governance_config.required_proposer_stake = required_proposer_stake;
+
+        let events = borrow_global_mut<GovernanceEvents>(@AptosFramework);
+        Event::emit_event<UpdateConfigEvent>(
+            &mut events.update_config_events,
+            UpdateConfigEvent {
+                min_voting_threshold,
+                required_proposer_stake,
+                voting_period_secs
+            },
+        );
+    }
+
+    /// Create a proposal with the backing `stake_pool`.
+    /// @param execution_hash Required. This is the hash of the resolution script. When the proposal is resolved,
+    /// only the exact script with matching hash can be successfully executed.
+    public(script) fun create_proposal(
+        proposer: &signer,
+        stake_pool: address,
+        execution_hash: vector<u8>,
+    ) acquires GovernanceConfig, GovernanceEvents {
+        let proposer_address = Signer::address_of(proposer);
+        assert!(Stake::is_delegated_voter(stake_pool, proposer_address), Errors::invalid_argument(ENOT_DELEGATED_VOTER));
+
+        // The proposer's stake needs to be at least the required bond amount.
+        let governance_config = borrow_global<GovernanceConfig>(@AptosFramework);
+        // TODO: Remove u128 cast once Coin is migrated to use u128 everywhere.
+        let stake_balance = (Stake::get_active_staked_balance(stake_pool) as u128);
+        assert!(
+            stake_balance >= governance_config.required_proposer_stake,
+            Errors::invalid_argument(EINSUFFICIENT_PROPOSER_STAKE),
+        );
+
+        // The proposer's stake needs to be locked up at least as long as the proposal's voting period.
+        let current_time = Timestamp::now_seconds();
+        let proposal_expiration = current_time + governance_config.voting_period_secs;
+        assert!(
+            Stake::get_lockup_secs(stake_pool) >= proposal_expiration,
+            Errors::invalid_argument(EINSUFFICIENT_STAKE_LOCKUP),
+        );
+
+        // We want to allow early resolution of proposals if more than 50% of the total supply of the network coins
+        // has voted. This doesn't take into subsequent inflation/deflation (rewards are issued every epoch and gas fees
+        // are burnt after every transaction), but inflation/delation is very unlikely to have a major impact on total
+        // supply during the voting period.
+        let total_voting_token_supply = Coin::supply<TestCoin>();
```

Comment:
similar question are all the coins in the staking pool?

Reply:
Nope but we can't just look at the current epoch's voting power as the voting period can span multiple epochs so voting power can change drastically. It's much safer to say that the proposal can only be resolved early if more than 50% actually voted yes or no (which implicitly means they must have staked > 50% of total supply)